src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -0,0 +1,362 @@
+	POCSFourierAmplitudeProj(IFourierMag,PFourier, lambda, radQuotient, (int)XSIZE(I()));
+	transformer.inverseFourierTransform();
+	POCSMinMaxProj(P(), Imin, Imax);
+	transformer.FourierTransform();
+	POCSFourierPhaseProj(PFourierPhase,PFourier);
+	transformer.inverseFourierTransform();
+ }
+
+ Image<double> ProgSubtractProjection::thresholdMask(Image<double> &m){
+ 	double maxMaskVol;
+ 	double minMaskVol;
+ 	m().computeDoubleMinMax(minMaskVol, maxMaskVol);
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(m())
+		DIRECT_MULTIDIM_ELEM(m(),n)=(std::abs(DIRECT_MULTIDIM_ELEM(m(),n)>maxMaskVol/20)) ? 1:0;
+	FilterG2.applyMaskSpace(m());
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(m())
+		DIRECT_MULTIDIM_ELEM(m(),n)=(std::abs(DIRECT_MULTIDIM_ELEM(m(),n)>0.5)) ? 1:0;
+	m.write(formatString("%smaskFocusThFilt.mrc", fnProj.c_str()));
+	for (int n=0; n<10; ++n) {
+		closing2D(m(), m(), 4, 0, 16);
+	}
+	m.write(formatString("%s/maskFocusThFiltClosing.mrc", fnProj.c_str()));
+ 	return m;
+ }
+
+ Image<double> ProgSubtractProjection::binarizeMask(Projection &m) const{
+ 	double maxMaskVol;
+ 	double minMaskVol;
+ 	m().computeDoubleMinMax(minMaskVol, maxMaskVol);
+ 	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(m())
+ 		DIRECT_MULTIDIM_ELEM(m(),n) =(DIRECT_MULTIDIM_ELEM(m(),n)>0.05*maxMaskVol) ? 1:0;
+ 	return m;
+ }
+
+ Image<double> invertMask(const Image<double> &m) {
+	Image<double> PmaskI = m;
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(PmaskI())
+		DIRECT_MULTIDIM_ELEM(PmaskI,n) = (DIRECT_MULTIDIM_ELEM(PmaskI,n)*(-1))+1;
+	return PmaskI;
+ }
+
+ Image<double> subtraction(Image<double> &I1, const Image<double> &I2,
+		 const Image<double> &minv, const Image<double> &m){
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(I1())
+		DIRECT_MULTIDIM_ELEM(I1,n) = (DIRECT_MULTIDIM_ELEM(I1,n)-(DIRECT_MULTIDIM_ELEM(I2,n)*DIRECT_MULTIDIM_ELEM(minv,n))) * DIRECT_MULTIDIM_ELEM(m,n);
+	return I1;
+ }
+
+ void ProgSubtractProjection::writeParticle(const int &ix, Image<double> &img) {
+	FileName out = formatString("%d@%s.mrcs", ix, fnOut.c_str());
+	img.write(out);
+	mdParticles.setValue(MDL_IMAGE, out, ix);
+ }
+
+ void ProgSubtractProjection::run() {
+	show();
+	V.read(fnVolR);
+	V().setXmippOrigin();
+ 	mdParticles.read(fnParticles);
+ 	maskVol = createMask(fnMaskVol, maskVol);
+ 	mask = createMask(fnMask, mask);
+	// Initialize Gaussian LPF to smooth mask
+	FilterG.FilterShape=REALGAUSSIAN;
+	FilterG.FilterBand=LOWPASS;
+	FilterG.w1=sigma;
+	// Initialize Gaussian LPF to threshold mask with ctf
+	FilterG2.FilterShape=REALGAUSSIAN;
+	FilterG2.FilterBand=LOWPASS;
+	FilterG2.w1=2;
+	// Initialize LPF to filter at desired resolution
+	Filter2.FilterBand=LOWPASS;
+	Filter2.FilterShape=RAISED_COSINE;
+	Filter2.raised_w=0.02;
+	Filter2.w1=cutFreq;
+	// Sizes for padding
+	pad = int(XSIZE(V())/2);
+    for (size_t i = 1; i <= mdParticles.size(); ++i) {
+    	row = mdParticles.getRowVec(i);
+    	readParticle(row);
+     	row.getValue(MDL_ANGLE_ROT, rot);
```

Comment:
either check that this label is present in the metadata, or use `getValueOrDefault()`